### PR TITLE
Feature/split component features

### LINKED ISSUE #421: Split up base component features
## Are you reporting a bug or a feature?

 - [x] Feature

## Expected Behavior

I would like to pick what my component needs, like:
- with lifecycle
- with context
- with template / incremental rendering
- with first class props / maybe state

## Actual Behavior

All is implemented together in `base-component`

### PR DESCRIPTION
**Pull request by AndyOGo**
Fixes #421 

Thanks to @TheDadi for the great idea:)

Changes proposed in this pull request:

 - rename `batchProps` to `setProps`
 - extracting base features
 - join constructor's multiple args to single object hash
 - adds a TOC to the contribution file
 - adds composition helper

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
 
 # How Has This Been Tested?
 
 
 
 # Checklist:
 
 - [x] My code follows the style guidelines of this project
 - [x] I have performed a self-review of my own code
 - [x] I have commented my code, particularly in hard-to-understand areas
 - [x] I have made corresponding changes to the documentation
 - [x] My changes generate no new warnings
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by AndyOGo** [comment]
@TheDadi 
@LucaMele 
We should settle on how to organise those HOCs - most are straight forward and easy, though other share common usage or need same callbacks...

**Straight forward**
- [ ] `withContext` should handle:
   - `conntextCallback` check inside the `constructor`
   - clean up inside `disconnectedCallback`
   - and all context-related methods as: `enableContext`, `selectContext`, `_makeContextReady` and `contextNode` getter
- [ ] `withUtils` should handle static `uuidv4`
- [ ] `withMonkeyPatches` should handle:
   - monkey patch of `innerText`, `textContent`, `innerHTML` setters and `appendChild` method

**Rather complex**
- [ ] `withBase` should setup common stuff, like:
   - `_isConnected` state (needed for incremental rendering only, but is very general)
   - `_hasTemplate` state (needed by monkey patches and incremental rendering)
   - `_initialise` method
- [ ] `withStyles` / `withGlobalStyles` should handle:
   - shared `_initiale` method
   - `_appendStyles` method
- [ ] `withLifecycle` should
   - may add the lifecycle logger
   - either just define the no-ops for `willRenderCallback` and `didRenderCallback` or nothing
- [ ] `withUpdate` should contain the whole incremental rendering logic, including
   - constructor stuff - adding `oberservedAttributes` setters, ...
   - `shouldUpdateCallback`
   - `attributeChangedCallback`, `batchProps`, `_reduceProps`
   - `render` method
   - `_isMorphing` state

File: src/js/abstract/base-component.js [hunk 1]
```diff
@@ -1,596 +1,73 @@
...
+     *
+     * @param  {type} styles description
+     * @param  {type} template description
+     * @return {type}        description
+     */
+    _initialise(styles, template = null) {
+      this._styles = styles;
+      this._template = template;
+      this._hasTemplate = !!template;
+      this._hasRendered = false;
```
Comment (AndyOGo) [inline src/js/abstract/base-component.js]: `this._hasRendered` belongs only to `withRender` and `withMonkeyPatches`

File: src/js/abstract/base-component.js [hunk 2]
```diff
@@ -1,596 +1,73 @@
...
-   * @returns {ContextNode|Boolean} - Returns an associated context node if found, else `false`.
-   */
-  get contextNode() {
-    const { __selectedContext } = this;
-    let { parentNode } = this;
-
-    while (parentNode && (!parentNode.__isContext || (__selectedContext && __selectedContext !== parentNode.__contextName))) {
-      // eslint-disable-next-line prefer-destructuring
-      parentNode = parentNode.parentNode;
+    static uuidv4() {
```
Comment (AndyOGo) [inline src/js/abstract/base-component.js]: @LucaMele 
Are we using this?
Should it be provided by it's own HOC?

File: src/js/abstract/hocs/with-context.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      if (this.unContextEnabled) {
+        this.unContextEnabled();
+      }
+    }
+
+    // @TODO: atm no data can be shared by enabling context, though this could be necessary
+    /**
+     * Provides an opt-in contextual scope for hierarchy-agnostic child components.
+     */
+    enableContext() {
```
Comment (AndyOGo) [inline src/js/abstract/hocs/with-context.js]: @LucaMele 
@TheDadi 
How about keeping the naming consistent?
Changing `enableContext` to `provideContext` and it's associated event either as `context/available` or `context/provided`?

File: src/js/abstract/hocs/with-context.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      // publish context/enabled with contextual node name
+      publish('context/enabled', contextName);
+    }
+
+    /**
+     * Opt-in to select a specific context by component name.
+     *
+     * @param {String} name - `nodeName` of the desired contextual node.
+     */
+    selectContext(name) {
```
Comment (AndyOGo) [inline src/js/abstract/hocs/with-context.js]: @LucaMele 
@TheDadi 
Same for `selectContext`, may change to `consumeContext`?

File: src/js/abstract/hocs/with-render.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+   *   </article>
+   * </axa-example>
+   * ```
+   */
+  class Render extends Base {
+    constructor(...args) {
+      super(...args);
+
+      // hook into updated
+      this.updated = this.render;
```
Comment (AndyOGo) [inline src/js/abstract/hocs/with-render.js]: @LucaMele 
@TheDadi 
`this.updated` is called by `withUpdate`.
Atm I have only this single function, though could be considered to change it to an event emitter (though no use case so far)

File: src/js/abstract/hocs/with-context.js (hunk 3 above)
Comment (LucaMele) [reply src/js/abstract/hocs/with-context.js]: naming is better like that yes

File: src/js/abstract/hocs/with-context.js (hunk 4 above)
Comment (LucaMele) [reply src/js/abstract/hocs/with-context.js]: naming is better like that yes

File: src/js/abstract/hocs/with-render.js (hunk 5 above)
Comment (LucaMele) [reply src/js/abstract/hocs/with-render.js]: then lets keep it for the moment

File: src/js/abstract/hocs/with-update.js [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import debounce from '../../debounce';
+import getAttribute from '../../get-attribute';
+
+const PROPERTY_WHITELIST = ['title', 'checked', 'disabled'];
+
+const withUpdate = Base =>
+  /**
+   * Adds attribute observation and enables **First Class Props**.
+   */
+  class Update extends Base {
```
[the remark this answers is not in the document]
Comment (LucaMele) [reply src/js/abstract/hocs/with-update.js]: instead of Update to be called WithUpdate? yes better

**Comment by AndyOGo** [comment]
@LucaMele 
No all components are fine as is

**Comment by LucaMele** [comment]
@AndyOGo 
i mean something like that
`const withAll = withContext(withMonkeyPatches(withUpdate(withRender(withStyles))))`

and then 

`withAll(...)`

File: src/js/abstract/base-component.js [hunk 7]
```diff
@@ -7,67 +7,67 @@ import withUpdate from './hocs/with-update';
...
+}
 
-    static uuidv4() {
-      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
-        let r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8); // eslint-disable-line
-        return v.toString(16);
-      });
-    }
-  })))));
+export default withContext(withMonkeyPatches(withUpdate(withRender(withStyles(BaseComponent)))));
```
Comment (LucaMele) [inline src/js/abstract/base-component.js]: or like this is perfect as well

File: src/js/abstract/hocs/with-shadow.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+const withShadow = Base =>
+  class WithShadow extends Base {
+    /**
+     * Attach shadow upon connect.
+     *
+     * @param {String} [mode='open']
+     */
+    connectedCallback(mode = 'open') {
+      const shadowRoot = this.attachShadow({ mode });
```
Comment (LucaMele) [inline src/js/abstract/hocs/with-shadow.js]: this code is not wotking

File: src/js/abstract/hocs/with-shadow.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+     *
+     * @param {String} [mode='open']
+     */
+    connectedCallback(mode = 'open') {
+      if (super.connectedCallback) {
+        super.connectedCallback();
+      }
+
+      const shadowRoot = this.attachShadow({ mode });
+      this._appendStyles(shadowRoot);
```
Comment (AndyOGo) [inline src/js/abstract/hocs/with-shadow.js]: @LucaMele 
How do we deal with appending styles for shadow DOM.
May styles should be an utility 🤔

**Comment by AndyOGo** [comment]
@LucaMele 
No worries. I got your point.
I just needed to swap it around by hoisting the the hoced class upwards.

File: src/js/abstract/hocs/with-shadow.js (hunk 9 above)
Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: this is a bit of legacy code. We need to call the append style same way as `BaseComponent` and NOT like `BaseComponentGlobal`. Then styles will be automatically scoped inside the shadow dom

Comment (AndyOGo) [reply src/js/abstract/hocs/with-shadow.js]: aha, let's fix it then

Comment (AndyOGo) [reply src/js/abstract/hocs/with-shadow.js]: is `withStyles` obsolete then?

Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: i think so. It should do what base component does before becoming base component global

Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: this lines are important

const shadowRoot = this.attachShadow({ mode });
this._appendStyles(shadowRoot);

Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: i dont know now where the logic is that was in that function

Comment (AndyOGo) [reply src/js/abstract/hocs/with-shadow.js]: okay, let's check it together.
We will make it work like magic 💪

File: src/js/abstract/hocs/with-shadow.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+const withShadow = Base =>
+  class WithShadow extends Base {
+    /**
+     * Attach shadow DOM upon connect.
+     *
+     * @param {String} [mode='open']
+     */
+    connectedCallback(mode = 'open') {
```
Comment (AndyOGo) [inline src/js/abstract/hocs/with-shadow.js]: I didn't know that `connectedCallback` receives a `mode` arg.
Where is that specified?

Comment (AndyOGo) [reply src/js/abstract/hocs/with-shadow.js]: According to these shadow DOM examples, attachShadow is called in the constructor.
https://developers.google.com/web/fundamentals/web-components/shadowdom

Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: I cant remeber what the idea was there to be honest..

Comment (LucaMele) [reply src/js/abstract/hocs/with-shadow.js]: my code comes also from a time where it was not standard yet

Comment (AndyOGo) [reply src/js/abstract/hocs/with-shadow.js]: okidok

File: src/js/abstract/base-component.js [hunk 11]
```diff
@@ -1,121 +1,22 @@
...
- * <axa-example>
- *   <article>
- *     <div>This is some light DOM for axa-example</div> <!-- light DOM injection point -->
- *   </article>
- * </axa-example>
- * ```
  */
-export default class BaseComponent extends HTMLElement {
+class BaseComponent extends HTMLElement {
   constructor(styles = '', template) {
```
Comment (AndyOGo) [inline src/js/abstract/base-component.js]: @LucaMele 
I would like to change the constructor to accept only a single arg of type `object`.
That we can forward this hash to each `super({})` call, and releated hoc can just destruct what they need 👍

**Comment by AndyOGo** [comment]
@LucaMele pointed out a very interesting aspect of composition being factorial and unidirectional, see #456

**Pull request merged**